lib/providers/user.dart

```diff
@@ -1,9 +1,46 @@
+import 'dart:async';
+import 'dart:convert';
 import 'package:flutter/material.dart';
+import '../util.dart';
 
 class User with ChangeNotifier {
   String _idToken; // Firebase ID token of the account.
   String _localId; // The uid of the current user.
   String email;
   String displayName;
   String photoUrl;
+
+  Future<Map<String, dynamic>> getDeviceIP() async {
```

Comment:
why getDeviceIP() returns Future<Map<String, dynamic>> ???
getDeviceIP() must return **"device's IP address"** which is just a simple String